### PR TITLE
Add ProofOfAddress .NET sample

### PR DESCRIPTION
**Pull request by fbouteruche**
Description of changes:

Adding the .NET sample which is the companion for the blog post https://medium.com/i-love-my-local-farmer-engineering-blog/blazor-webassembly-applications-7cfece9609f6

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by awsdb — CHANGES_REQUESTED** [review]
Unfortunately, the CDK deployment fails because the generated Cognito domain is not globally unique. This should be fixed before publishing.

Additionally, I posted a couple of comments about the code in general. Feel free to address those items if you like my suggestions.

File: ProofOfAddress/src/Web/Shared/LoginDisplay.razor [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    private Task<AuthenticationState>? authenticationStateTask { get; set; }
+
+    private string? _userName;
+
+    protected override async void OnParametersSet()
+    {
+        base.OnParametersSet();
+        var authState = authenticationStateTask is not null ? await authenticationStateTask : null;
+        var user = authState is not null ? authState.User : null;
+        if(user is not null && user.Identity is not null && user.Identity.IsAuthenticated)
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Web/Shared/LoginDisplay.razor:28]: There are some debates here but your version is the way to go nowadays :)

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+            {
+                AssumedBy = new ServicePrincipal("build.apprunner.amazonaws.com"),
+                ManagedPolicies = new[]
+                {
+                    ManagedPolicy.FromAwsManagedPolicyName("service-role/AWSAppRunnerServicePolicyForECRAccess")
+                }
+            });
+
+            // create an AWS App Runner service using the container image published to Amazon ECR and the IAM role
+            var appRunnerService = new CfnService(this, "ProofOfAddressApi", new CfnServiceProps
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:116]: Add a section about which region to use.

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+                    }
+                }
+            });
+
+            // Create a User Pool domain so that user can log in throught the Amazon Cognito Hosted UI
+            var userPoolDomain = userPool.AddDomain("UserPoolDomain", new UserPoolDomainOptions()
+            {
+                CognitoDomain = new CognitoDomainOptions()
+                {
+                    DomainPrefix = Names.UniqueResourceName(userPool, new UniqueResourceNameOptions()).ToLower()
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs]: used a GUID

File: ProofOfAddress/src/API/Controllers/AuthController.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                currentUser = new CurrentUser(false, String.Empty, new Dictionary<string, string>());
+            }
+            else
+            {
+                Dictionary<string, string> claims = new Dictionary<string, string>();
+                foreach (var claim in this.User.Claims)
+                {
+                    claims.Add(claim.Type, claim.Value);
+                }
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/API/Controllers/AuthController.cs]: integrated suggestion

File: ProofOfAddress/src/API/Controllers/PresignedController.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
+﻿using Amazon.CognitoIdentity;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authorization;
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/API/Controllers/PresignedController.cs]: integrated suggestion

File: ProofOfAddress/src/API/Controllers/PresignedController.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            _configuration = configuration;
+            _logger = logger;
+        }
+
+        //[Authorize]
+        public async Task<string> Get()
+        {
+            _logger.LogInformation("Start processing get presigned request");
+
+            // Retrieve the ID token from the HttpContexte and use it to get user temporary credentials from Amazon Cognito
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/API/Controllers/PresignedController.cs]: Fix typo

File: ProofOfAddress/src/Web/Config/RemoteConfigurationProvider.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+﻿using Microsoft.Extensions.Primitives;
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Web/Config/RemoteConfigurationProvider.cs]: Integrated suggestion

File: ProofOfAddress/src/Web/Providers/BFFAuthenticationStateProvider.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+        public BFFAuthenticationStateProvider(ICurrentUserService currentUserService, ILogger<BFFAuthenticationStateProvider> logger)
+        {
+            _currentUserService = currentUserService;
+            _logger = logger;
+        }
+
+        public override async Task<AuthenticationState> GetAuthenticationStateAsync()
+        {
+            var identiy = new ClaimsIdentity();
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Web/Providers/BFFAuthenticationStateProvider.cs]: Fix typo

File: ProofOfAddress/src/Web/Providers/BFFAuthenticationStateProvider.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                var userInfo = await GetCurrentUser();
+                if (userInfo.IsAuthenticated)
+                {
+                    var claims = new List<Claim>();
+                    claims.Add(new Claim(ClaimTypes.Name, userInfo!.UserName));
+                    foreach (var claim in userInfo.Claims)
+                    {
+                        claims.Add(new Claim(claim.Key, claim.Value));
+                    }
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Web/Providers/BFFAuthenticationStateProvider.cs]: Integrated suggestion

File: ProofOfAddress/src/Web/Shared/LoginDisplay.razor (hunk 1 above)
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Web/Shared/LoginDisplay.razor]: Integrated suggestion

File: ProofOfAddress/README.md [hunk 10]
```diff
@@ -56,7 +62,7 @@ On Linux, if your docker daemon runs with privileges, you may need to prefix `cd
 To build and deploy the sample, run the following commands:
 
 ```bash
-git clone https://github.com/aws-samples/i-love-my-local-farmer
+git clone https://gitlab.aws.dev/i-love-my-local-farmer/proof-of-address.git
```
Comment (balabad) [inline ProofOfAddress/README.md:65]: Are you sure this URL is correct? This is our internal GitLab instance and it is not publicly available.

File: ProofOfAddress/src/Web/wwwroot/sample-data/weather.json [hunk 11]
```diff
@@ -0,0 +1,27 @@
+﻿[
```
Comment (balabad) [inline ProofOfAddress/src/Web/wwwroot/sample-data/weather.json:1]: Are you sure we need this file?

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs]: Revert the solution as using a GUID would change the domain at each update. Use the concatenation of Account ID and Unique Resource Name.

File: ProofOfAddress/src/API/Controllers/PresignedController.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            var idToken = await HttpContext.GetTokenAsync("Cognito", "id_token");
+            CognitoAWSCredentials credentials = new CognitoAWSCredentials(_configuration["IdentityPoolId"], Amazon.RegionEndpoint.GetBySystemName(_configuration["IdentityPoolRegion"]));
+            _logger.LogInformation($"Authority: {_configuration["Authority"]}");
+            _logger.LogInformation($"TokenProviderName: {_configuration["Authority"].Replace("https://", "")}");
+            credentials.AddLogin(_configuration["Authority"].Replace("https://", ""), idToken);
+            await credentials.GetCredentialsAsync();
+
+            // Use the user temporary credentials to get a presigned URL from Amazon S3
+            return GeneratePresignedUrl(credentials, 600, _configuration["BucketName"], $"{Guid.NewGuid()}.jpg");
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/API/Controllers/PresignedController.cs:36]: Done

File: ProofOfAddress/src/API/Controllers/PresignedController.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        private string GeneratePresignedUrl(CognitoAWSCredentials credentials, int duration, string bucket, string key)
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/API/Controllers/PresignedController.cs:39]: Fixed

File: ProofOfAddress/src/ConfigFunction/Config.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/ConfigFunction/Config.cs:5]: Fixed

File: ProofOfAddress/src/ConfigFunction/Function.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+    return new
+    APIGatewayProxyResponse()
+    {
+        StatusCode = 200,
+        Body = System.Text.Json.JsonSerializer.Serialize(new {
+            GetPresignedUrl = Environment.GetEnvironmentVariable(MyLocalFarmer.ProofOfAddress.ConfigFunction.Config.GET_PRESIGNED_URL.ToString()),
+            LoginUrl = Environment.GetEnvironmentVariable(MyLocalFarmer.ProofOfAddress.ConfigFunction.Config.LOGIN_URL.ToString()),
+            LogoutUrl = Environment.GetEnvironmentVariable(MyLocalFarmer.ProofOfAddress.ConfigFunction.Config.LOGOUT_URL.ToString()),
+            GetCurrentUserUrl = Environment.GetEnvironmentVariable(MyLocalFarmer.ProofOfAddress.ConfigFunction.Config.GET_CURRENT_USER_URL.ToString())
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/ConfigFunction/Function.cs:18]: Fixed

File: ProofOfAddress/src/Infra/MyLocalFarmer.ProofOfAddress.Infra.csproj [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+  <ItemGroup>
+    <!-- CDK Construct Library dependencies -->
+    <PackageReference Include="Amazon.CDK.AWS.Cognito.IdentityPool.Alpha" Version="2.33.0-alpha.0" />
+    <PackageReference Include="Amazon.CDK.Lib" Version="2.33.0" />
+    <PackageReference Include="Constructs" Version="10.1.58" />
+
+    <!-- jsii Roslyn analyzers (un-comment to obtain compile-time checks for missing required props
+    <PackageReference Include="Amazon.Jsii.Analyzers" Version="*" PrivateAssets="all" />
+    -->
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmer.ProofOfAddress.Infra.csproj:24]: Done

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+using Amazon.CDK.AWS.AppRunner;
+using Amazon.CDK.AWS.CloudFront;
+using Amazon.CDK.AWS.CloudFront.Origins;
+using Amazon.CDK.AWS.Cognito;
+using Amazon.CDK.AWS.Cognito.IdentityPool.Alpha;
+using Amazon.CDK.AWS.Ecr.Assets;
+using Amazon.CDK.AWS.IAM;
+using Amazon.CDK.AWS.Lambda;
+using Amazon.CDK.AWS.S3;
+using Amazon.CDK.AWS.S3.Assets;
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:12]: Fixed

File: ProofOfAddress/src/Infra/GlobalSuppressions.cs [hunk 18]
```diff
@@ -0,0 +1 @@
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Potential Code Quality Issues", "RECS0026:Possible unassigned object created by 'new'", Justification = "Constructs add themselves to the scope in which they are created")]
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/GlobalSuppressions.cs:1]: The rule CA1806 is not yeld

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+            var fileStorage = new Bucket(this, "FileStorage", new BucketProps
+            {
+                AutoDeleteObjects = true,
+                RemovalPolicy = RemovalPolicy.DESTROY,
+                BlockPublicAccess = BlockPublicAccess.BLOCK_ALL,
+                Cors = new CorsRule[]
+                {
+                    new CorsRule()
+                    {
+                        AllowedHeaders = new string[]{"*"},
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:245]: Fixed

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 20]
```diff
@@ -0,0 +1,437 @@
...
+                        AllowedMethods = new HttpMethods[]{HttpMethods.PUT},
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:246]: Fixed

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+                        AllowedOrigins = new string[]{$"https://{cloudFrontDistribution.DomainName}"},
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:247]: Fixed

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 22]
```diff
@@ -0,0 +1,437 @@
...
+                        ExposedHeaders = new string[]{},
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:248]: Fixed

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+                    }
+                }
+            });
+            // Grant write permissions to authenticated users from the Amazon Cognito Identity Pool
+            fileStorage.GrantWrite(identityPool.AuthenticatedRole);
+            #endregion
+
+            #region UPDATE AWS APP RUNNER SERVICE ENVIRONMENT VARIABLE
+            var appRunnerCustomResource = new AwsCustomResource(this, "AppRunnerServiceEnvironmentVariables", new AwsCustomResourceProps()
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:257]: Fixed

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+                PriceClass = PriceClass.PRICE_CLASS_100
+            });
+
+
+            // Bundle the single-page application content and upload it to the contentS3Bucket
+            IEnumerable<string> publishCommands = new[]
+            {
+                "export DOTNET_CLI_HOME=\"/tmp/DOTNET_CLI_HOME\"",
+                "export PATH=\"$PATH:/tmp/DOTNET_CLI_HOME/.dotnet/tools\"",
+                $"cd {nameof(MyLocalFarmer.ProofOfAddress.Web)}",
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:58]: It is on purpose. The idea here is to break the build if you refactor your solution structure and namespaces. It is a weak defense.

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 25]
```diff
@@ -0,0 +1,437 @@
...
+                Handler = String.Join(".", nameof(MyLocalFarmer), nameof(ProofOfAddress), nameof(ConfigFunction))
+            });
+
+            LambdaRestApi configApi = new LambdaRestApi(this, "ConfigApi", new LambdaRestApiProps()
+            {
+                Handler = configFunction,
+            });
+
+            cloudFrontDistribution.AddBehavior("config", new RestApiOrigin(configApi), new BehaviorOptions()
+            {
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:375]: It is a wrong alert

[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:375]: The alert is valid when you create a new Distribution object. Here, I'm adding a new behavior to the distribution. The origin is define by the second argument of the AddBehavior method, new RestApiOrigin(configApi) in this case.

File: ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs [hunk 26]
```diff
@@ -0,0 +1,437 @@
...
+                    Service = "Lambda",
+                    Action = "updateFunctionConfiguration",
+                    Parameters = new Dictionary<string, object>
+                    {
+                        { "FunctionName", configFunction.FunctionArn },
+                        { "Environment", new Dictionary<string, object>()
+                            {
+                                { "Variables", new Dictionary<string, string>()
+                                    {
+                                        { MyLocalFarmer.ProofOfAddress.ConfigFunction.Config.GET_PRESIGNED_URL.ToString(), $"https://{appRunnerService.AttrServiceUrl}/presigned" },
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/MyLocalFarmerProofOfAddressInfraStack.cs:398]: Fixed

File: ProofOfAddress/src/Infra/Program.cs [hunk 27]
```diff
@@ -0,0 +1,44 @@
+﻿using Amazon.CDK;
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/Program.cs:4]: Fixed

File: ProofOfAddress/src/Web/wwwroot/sample-data/weather.json (hunk 11 above)
Comment (fbouteruche) [reply ProofOfAddress/src/Web/wwwroot/sample-data/weather.json]: Deleted

File: ProofOfAddress/README.md (hunk 10 above)
Comment (fbouteruche) [reply ProofOfAddress/README.md]: Fixed

File: ProofOfAddress/src/Shared/CurrentUser.cs [hunk 28]
```diff
@@ -0,0 +1,23 @@
+﻿namespace MyLocalFarmer.ProofOfAddress.Shared
+{
+    public class CurrentUser
+    {
+        public bool IsAuthenticated { get; set; }
+        public string UserName { get; set; }
+        public IDictionary<string, string> Claims { get; set; }
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Shared/CurrentUser.cs]: Transform to record

File: ProofOfAddress/src/Infra/README.md [hunk 29]
```diff
@@ -0,0 +1,14 @@
+# Welcome to your CDK C# project!
```
[the remark this answers is not in the document]
Comment (fbouteruche) [reply ProofOfAddress/src/Infra/README.md]: Fixed

**Review by balabad — APPROVED** [review]

**Pull request merged**